### ISSUE TITLE
Skyrider glitches out, claims to be going back to Drassen Airport but doesn't.

### ISSUE THREAD

**Issue by Ilyab99** [ja2-stracciatella/ja2-stracciatella#554]

**Comment by Ilyab99** [comment]
Can't upload a .sav for some reason 

OneDrive link https://1drv.ms/u/s!AkQU7CFgL4Wwh2Z-Iu52p4lQbvDS

**Comment by Manuel-K** [comment]
You can upload it if you compress it in a zip file.

**Comment by Ilyab99** [comment]
It isn't working for some reason. It says "We don’t support that file type.  Choose Files Try again with a PNG, GIF, JPG, DOCX, PPTX, XLSX, TXT, PDF, or ZIP."

**Comment by lynxlynxlynx** [comment]
this could be related to a past cleanup @derSteFfi did, so a bisect run could be the fastest way to find when the problems started.

**Comment by Ilyab99** [comment]
What's a bisect run?

**Comment by lynxlynxlynx** [comment]
testing with "git bisect". If you compiled ja2s yourself, you could give it a try. It goes basically like this (unless the build breaks):
git bisect start
git bisect bad # curent state is buggy
git checkout HEAD~500 # switch to older code
build it
test it
if it works: git bisect good # then git will lead you forward, eventually narrowing down to the problematic commit
if it doesn't, checkout even older code.

**Comment by Ilyab99** [comment]
Alright, I'm going to take a look at the code (maybe gain an understanding of what does what) and see if i can pinpoint an error. I'll also do the bisect thing.

**Comment by derSteFfi** [comment]
i tried to load the savegame with commit ed48c1f283f76a1ea475989a064726a39a5aa7f5, which is right before i started the helicopter cleanup, didn't help. Maybe Skyrider somehow got into an awkward state in the savegame due to the cleanup, but I honestly could not see how. I'going to dig deeper, too.

**Comment by JAsmine-ja2** [comment]
I loaded the savegame in vanilla JA2 1.12 Gold English and in Stracciatella 0.15.1 and can confirm the bug in both versions.

In vanilla JA2, things get quite buggy. First thing, the airspace map looks like this:
![screen000](https://user-images.githubusercontent.com/17213795/32417323-16dc2d2c-c258-11e7-9e44-8c5e104a9ce6.jpg)

And if I play around with this savegame some more, I get to a point where I can neither save the game or reload the savegame before restarting the game...

For reference and to prevent it getting lost, here is the savegame from Ilyab99:
[QuickSave.zip](https://github.com/ja2-stracciatella/ja2-stracciatella/files/1444261/QuickSave.zip)

**Comment by chmallon** [comment]
Here is some debug info about the movement group of the helicopter:
> BAD group 50 in G9 has next waypoint G11 but does not move
> INFO persistent player group 50, size 0, sector G9, next <NOWHERE>, prev G8, next waypoint: 0, waypoints: G11 E11 E12 D12 D13 B13

So there is a list of waypoints to Drassen airport, but the group is currently not moving.
Also the flag 'fHeliReturnStraightToBase' is set.

BTW: The five boolean variables to model the state of the helicopter, which theoretically represent 32 states, can be condensed into one variable with just 6 states.

**Comment by derSteFfi** [comment]
Yeah, the Helicopter is abeast of source code. I did a first clean up a while back trying to not change anything semantically. There's still much room for improvement.

If the group has waypoints but is not moving, moving wasn't started for some reason. I'll see if I can find it.

**Comment by chmallon** [comment]
The whole movement code is a mess.
E.g. each soldier has its own copy of the movement path, but all soldiers in a squad must move the same way and soldiers can only move in a squad, so one path per squad would be enough.
This duplication must be carefully synchronized and of course there are vanilla bugs, e.g. I fixed one in 0.9 which caused that soldiers in one squad would get different destionations.
Further, the PathSt entries are doubly-linked for no good reason, the functions to manipulate them are extremely convoluted, movement groups are pre-allocated for vehicles for unclear reasons, while the slots for the vehicles themselves are allocated on demand, vehicles always get added to the end of the team list for unclear reaons, there's a lot of error-prone special case handling for Skyrider (sometimes misspelled as Skyridder) and the helicopter, ...
